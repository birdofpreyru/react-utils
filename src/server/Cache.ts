type CachedItemT<DatumT> = {
  data: DatumT;
  size: number;
  timestamp: number;
};

/**
 * Implements the static cache.
 */
export default class Cache<DatumT> {
  private items: Record<string, CachedItemT<DatumT>> = {};

  private size = 0;

  constructor(private maxSize: number) { }

  /**
   * Cache lookup.
   * @param key Item key to look for.
   * @param [maxage=Number.MAX_VALUE] Optional. The maximum age of
   *  cached item to serve. Default to infinite.
   * @returns Cached item, or null if the item is absent in cache,
   *  or stale.
   */
  get({
    key,
    maxage = Number.MAX_VALUE,
  }: {
    key: string;
    maxage?: number;
  }): DatumT | null {
    const item = this.items[key];
    return item && Date.now() - item.timestamp < maxage ? item.data : null;
  }

  /**
   * Adds item to cache.
   * @param data Item to add.
   * @param key Key to store the item at.
   * @param size Byte size of the item.
   */
  add(data: DatumT, key: string, size: number): void {
    const cached = this.items[key];
    if (cached) this.size -= cached.size;

    this.items[key] = { data, size, timestamp: Date.now() };
    this.size += size;

    if (this.size > this.maxSize) {
      const entries = Object.entries(this.items);
      entries.sort((a, b) => a[1].timestamp - b[1].timestamp);

      for (const entry of entries) {
        const [itemKey, item] = entry;
        delete this.items[itemKey];
        this.size -= item.size;
        if (this.size < this.maxSize / 2) break;
      }
    }
  }
}
