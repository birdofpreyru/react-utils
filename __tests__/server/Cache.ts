/**
 * Testing Cache object.
 */

import mockdate from 'mockdate';
import Cache from 'server/Cache';

mockdate.set('2021-04-05Z');

it('works as expected', () => {
  const cache = new Cache(18);
  expect(cache).toMatchSnapshot();
  cache.add('123', 'A', 2 * 3);
  mockdate.set(Date.now() + 1000);
  cache.add('123456', 'B', 2 * 6);
  mockdate.set(Date.now() + 1000);
  expect(cache).toMatchSnapshot();
  cache.add('ab', 'A', 2 * 2);
  mockdate.set(Date.now() + 1000);
  expect(cache).toMatchSnapshot();
  expect(cache.get({ key: 'C' })).toBeNull();
  expect(cache.get({ key: 'A' })).toBe('ab');
  expect(cache.get({ key: 'A', maxage: 0 })).toBeNull();
  cache.add('21', 'C', 2 * 2);
  expect(cache).toMatchSnapshot();
});
