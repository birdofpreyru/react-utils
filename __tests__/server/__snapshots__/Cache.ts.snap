// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`works as expected 1`] = `
Cache {
  "items": {},
  "maxSize": 18,
  "size": 0,
}
`;

exports[`works as expected 2`] = `
Cache {
  "items": {
    "A": {
      "data": "123",
      "size": 6,
      "timestamp": 1617580800000,
    },
    "B": {
      "data": "123456",
      "size": 12,
      "timestamp": 1617580801000,
    },
  },
  "maxSize": 18,
  "size": 18,
}
`;

exports[`works as expected 3`] = `
Cache {
  "items": {
    "A": {
      "data": "ab",
      "size": 4,
      "timestamp": 1617580802000,
    },
    "B": {
      "data": "123456",
      "size": 12,
      "timestamp": 1617580801000,
    },
  },
  "maxSize": 18,
  "size": 16,
}
`;

exports[`works as expected 4`] = `
Cache {
  "items": {
    "A": {
      "data": "ab",
      "size": 4,
      "timestamp": 1617580802000,
    },
    "C": {
      "data": "21",
      "size": 4,
      "timestamp": 1617580803000,
    },
  },
  "maxSize": 18,
  "size": 8,
}
`;
